guides/api-guideline.md

```diff
@@ -30,6 +32,81 @@ The following basic data types should be used in the API documentation.
 | `dateTime`        | A date and time value without timezone (nanosecond precision) |
 | `zonedDateTime`   | A date and time value with timezone (nanosecond precision)    |
 
+### Complex Types
+
+Complex data types can be defined using the basic data types and other complex data types.
+A complex data type definition can contain [attributes](#attributes) and [methods](#methods).
+The following syntax should be used to define complex data types:
+```
+DataTypeName {
+    fieldName1: DataType1
+    fieldName2: DataType2
+    ReturnType methodName(param1: DataType1, param2: DataType2)
+}
+```
+
+#### Complex Type annotations
+
+Type annotations apply to a complex data type as a whole (as opposed to a single attribute or a method).
+The following annotations should be used:
+- `@@oneOf(field1, field2[, ...])`: Exactly one of the referenced fields can be non-null/non-undefined at any given time.
+- `@@oneOrNoneOf(field1, field2[, ...])`: Exactly one of the referenced fields can be non-null/non-undefined at any given time or all fields are null/undefined.
+
+Rules and recommendations for `@@oneOf`:
+- All listed fields must be declared with `@@nullable`
+- None or all listed fields must be declared with `@@immutable`.
+- If the fields are not `@@immutable` and a field is set, an SDK must unset all other fields
+- If the fields are not `@@immutable` and the currently set field is unset, an SDK must throw an error.
+- None or exactly one of the listed fields must be annotated by `@@default(value)`
+- If the fields are not `@@immutable` and none is annotated by `@@default(value)` at least one must be set by the constructor to not end in an invalid state. 
```

Comment:
Must it be set in the constructor, or at least set by the user before use? If set in the constructor, I think that would warrant a `@@default` tag